### PR TITLE
[clang-tidy] In C++17, callee is guaranteed to be sequenced before arguments.

### LINKED ISSUE #57758: clang-tidy 14.0.6 gives a false-positive bugprone-use-after-move report in chained expressions
Consider the following code
```cpp
#include <cassert>
#include <vector>

struct test {
 test& foo(std::vector<int> const& v) { assert(v.size() == 1); return *this; }
 int bar(std::vector<int> v ) { return v.size(); }
};

int main() {
    std::vector<int> v = { 42 };
    return test{}.foo(v).bar(std::move(v));

}
```

clang-tidy 14.0.6 reports the following error:

```console
tidy14-bug/main.cpp:11:23: error: 'v' used after it was moved [bugprone-use-after-move,-warnings-as-errors]
    return test{}.foo(v).bar(std::move(v));
                      ^
tidy14-bug/main.cpp:11:30: note: move occurred here
    return test{}.foo(v).bar(std::move(v));
                             ^
tidy14-bug/main.cpp:11:23: note: the use and move are unsequenced, i.e. there is no guarantee about the order in which they are evaluated
    return test{}.foo(v).bar(std::move(v));
                      ^
```

The standard, however, clearly [defines the behavior](http://eel.is/c++draft/expr.call#8) for the case.
This false-positive breaks a lot of code in our codebase.

### LINKED ISSUE #59612: Clang-tidy: bugprone-use-after-move: false negative
Reproduction:
````cpp
// https://compiler-explorer.com/z/8asKxxeGh
// clang++ -std=c++20  -stdlib=libc++  -Weverything -Wno-c++98-compat -Wno-missing-prototypes t.cpp
// clang-tidy -checks=bugprone-use-after-move
#include <utility>

struct C {};

struct S {
    auto g(C c) -> void;
};

auto f(C &c) -> S;

int main(int, char**) {
    auto c = C{};
    f(c).g(std::move(c));
}

````

Expected:
````console
[<source>:13:7: warning: 'c' used after it was moved [bugprone-use-after-move]]
    f(c).g(std::move(c));
      ^
[<source>:13:12: note: move occurred here]
    f(c).g(std::move(c));
           ^
[<source>:13:7: note: the use and move are unsequenced, i.e. there is no guarantee about the order in which they are evaluated]
    f(c).g(std::move(c));
      ^
1 warning generated.
````

When the argument of `f` is changed from `C &c` to either `const C &c` or `C c`, the expected warning shows up. However, when the class is an (in/)out argument, this does not show. I don't believe there is a difference in the sequence of execution for these cases, so the warning should happen in all cases.

### PR DESCRIPTION
**Pull request by tchaikov**
This eliminates false positives in bugprone-use-after-move where a variable
is used in the callee and moved from in the arguments.

We introduce one special case: If the callee is a MemberExpr with a DeclRefExpr as its base, we consider it to be sequenced after the arguments. This is because the variable referenced in the base will only actually be accessed when the call happens, i.e. once all of the arguments have been evaluated. This has no basis in the C++ standard, but it reflects actual behavior that is relevant to a use-after-move scenario:
```c++
a.bar(consumeA(std::move(a));
```
In this example, we end up accessing a after it has been moved from, even though nominally the callee a.bar is evaluated before the argument consumeA(std::move(a)).

Treating this scenario correctly has required rewriting the logic in bugprone-use-after-move that governs whether the use happens in a later loop iteration than the move. This was previously based on an unsound heuristic (does the use come lexically before the move?); we now use a more rigourous criterion based on reachability in the CFG.

Fixes #57758
Fixes #59612

### CODE REVIEW

**Comment by tchaikov** [comment]
@martinboehme hello Martin, I resurrected your change at https://reviews.llvm.org/D145581?id=503330#inline-1406063 and posted here in hope that we can continue your efforts and finally land the change in main branch. Hope you don't mind that I created this without your permission. we are also using clang-tidy in our project, since we are using the chained expression like `v.foo(x, y).then(std::move(x)).then(std::move(v))` a lot, it's a little bit annoying that clang-tidy warns at seeing this. this reduces the signal-to-noise ratio of its output. 

On top of your change, I made following changes

- rebase onto the latest main HEAD
- s/const auto &Succ/const CFGBlock *Succ/, so it is more explicit that the element type is a pointer.
- use `Visited.Contains(e)` instead of `!Visited.count(e)`, for better readability
- rename `found` to `Found`, to be more consistent with the naming convention in LLVM
- check for null before pushing a new successor node to `Stack`, otherwise we could be iterating a "null" node's successors.

these changes are collected in a separate commit in this PR, if you are good with them, I will fold it into the first commit.

**Comment by martinboehme** [comment]
> @martinboehme hello Martin, I resurrected your change at https://reviews.llvm.org/D145581?id=503330#inline-1406063 and posted here in hope that we can continue your efforts and finally land the change in the main branch. Hope you don't mind that I created this PR without your permissions.

No, not at all. On the contrary -- thanks for picking up this change that I let go stale.

> We are also using clang-tidy in our project. Since we are using the chained expression like `v.foo(x, y).then(std::move(x)).then(std::move(v))` a lot, it's a little bit annoying that clang-tidy warns at seeing this. These false alarms reduce the signal-to-noise ratio of the output of this tool.
> 
> On top of your change, I made following changes
> 
> [snip]
> 
> these changes are collected in a separate commit in this PR, if you are good with them, I will fold it into the first commit, and then mark this PR ready-for-review.

Yes, those changes look good to me.

Thanks!

**Comment by tchaikov** [comment]
@martinboehme thank you! from now on, i will try to address the upcoming comments from reviewers if this PR is fortunate enough to get more attentions, and to follow it up.

**Comment by tchaikov** [comment]
v2:

- s/auto *Member/const auto *Member/
- merge into the existing ReleaseNote entry of the same check, instead of creating another one

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:224]: agreed. but practically, we reference `EvaluationOrderUndefined` only if `UseAfterMoveFinder::find()` returns `true`, which indicates a use-after-move is identified. see https://github.com/llvm/llvm-project/blob/4d95850d052336a785651030eafa0b24651801a0/clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp#L498-L501, and the only two places where we return `true` in `UseAfterMoveFinder::findInternal()` are

1. https://github.com/llvm/llvm-project/blob/4d95850d052336a785651030eafa0b24651801a0/clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp#L178
2. https://github.com/llvm/llvm-project/blob/4d95850d052336a785651030eafa0b24651801a0/clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp#L188

the 2nd case is a recursive call to `UseAfterMoveFinder::findInternal()`. and the 1st case is where we set `UseHappensInLaterLoopIteration` to `false`. and we may set this member variable to `true` later on.

probably a more readable implementation is to return an `optional<UseAfterMove>` from `UseAfterMoveFinder::find()`, so that it's obvious that `UseAfterMove` is always initialized and returned if we find a use-after-move. but that'd be a cleanup, and not directly related to this PR. if it acceptable to piggy back it into this PR, i will do it.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:99]: yeah, it's equivalent to `reaches()`. will use it instead in the next revision.

File: clang-tools-extra/clang-tidy/utils/ExprSequence.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/utils/ExprSequence.cpp:74]: sure.

File: clang-tools-extra/clang-tidy/utils/ExprSequence.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/utils/ExprSequence.cpp:157]: ack.

File: clang-tools-extra/clang-tidy/utils/ExprSequence.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/utils/ExprSequence.cpp:133]: sure.

**Comment by tchaikov** [comment]
v3:

- trade `reaches()` helper for `CFGReverseBlockReachabilityAnalysis`. less repeating this way.
- replace `argsContain()` helper with `llvm::is_contained()`. less repeating this way.
- s/call/Call/. more consistent with the naming convention in LLVM.
- initialize `EvaluationOrderUndefined` in-class

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:202]: after a second thought, i am removing this initialization, and just set the default value in the `UseAfterMove`. and move the accompanied comment there.

**Comment by tchaikov** [comment]
@5chmidti Thanks for your thoughtful review and suggestions!
 I've incorporated them into the latest revision, which I'd appreciate you taking another look at.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:142]: `CFA` does not need to be a unique_ptr, you can simply create a `CFGReverseBlockReachabilityAnalysis` on the stack.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (5chmidti) [inline clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:39]: While you're at it, please initialize `EvaluationOrderUndefined` to `false` as well.

**Review by 5chmidti — APPROVED** [review]
LGTM after fixing these two comments, but let's wait for others to review as well

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:39]: sure. but for the reason explained at https://github.com/llvm/llvm-project/pull/93623#discussion_r1631614642. but please note, it's just for the sake of readability / consistency, not for the correctness.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp:142]: `CFGReverseBlockReachabilityAnalysis` comes with two member variables of `llvm::BitVector` and `llvm::DenseMap<>`, which could be potentially too large to put on the stack, i thought. that's why i allocated it on heap. but if you believe it's safe to do so. will allocate it on stack.

**Comment by tchaikov** [comment]
v3:

- allocate `CFGReverseBlockReachabilityAnalysis` on stack not on heap, as it's small enough and can be fit in the stack.
- initialize `EvaluationOrderUndefined` in-class to be more consistent. please note, before this change, it's always initialized if it's going to be referenced.

**Comment by tchaikov** [comment]
@5chmidti hi Julian, thank you for your review, suggestions and approval. updated accordingly.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (martinboehme) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp]: Drive-by comment: These types are fine to put on the stack. Note that the elements contained in these containers are stored on the heap in any case. For example, in the case of `DenseMap`, this storage is allocated by `DenseMap::allocateBuckets()`. The only things contained in `DenseMap` itself (and hence stored on the stack) are a pointer to the bucket array `Buckets` and some counters (`NumEntries`, `NumBuckets`, `NumTombstones`).

**Comment by tchaikov** [comment]
@PiotrZSL @SimplyDanny and @HerrCai0907 Hello, gentlemen. Would you be available to take a look at this at your earliest convenience?

**Review by PiotrZSL — APPROVED** [review]
LGTM.

**Comment by tchaikov** [comment]
@5chmidti @PiotrZSL Hi Julian and Piotr, thanks for taking the time to review my PR. I'm eager to get it merged. Is there anything else I can do to help facilitate the merge process? Or we can merge it now?

**Comment by tchaikov** [comment]
@5chmidti @PiotrZSL Hi Julian and Piotr, I hope this message finds you well. I'm following up on my previous comment regarding the PR I submitted two weeks ago. I understand you both might be busy, but I wanted to check if there's been any progress or if we are expecting more inputs from other reviewers to move forward with the merge.
If that's not the case, could you please help merge this change?

**Comment by HerrCai0907** [comment]
Normally PR can be merged if someone approved PR. You can merge it by yourself.
If you don't have access, I'm glad to help you.

**Comment by tchaikov** [comment]
@HerrCai0907 Hello Congcong, thank you for your message. I appreciate your willingness to help. As I don't have write access to the repository, I would be grateful if you could review and merge this pull request for me. Please let me know if you need any additional information or if you have any questions about the changes.

**Comment by tchaikov** [comment]
thank you Piotr for merging this change.

File: clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp
Comment (tchaikov) [reply clang-tools-extra/clang-tidy/bugprone/UseAfterMoveCheck.cpp]: > for better readability. but the `optional<>` refactor is still a nice-to-have, IMHO.

created #98100

**Pull request merged**
